basedtyping/reified_generic.py

```diff
@@ -0,0 +1,160 @@
+            raise UnboundTypeVarError(
+                f"generic alias with unbound generics detected: {self.__parameters__}"
+            )
+
+    def _type_vars(self) -> tuple[TypeVar, ...]:
+        """gets a ``tuple`` of all the ``TypeVar``s defined in the `__origin__`.
+
+        basically you should always use this instead of ``self.__parameters__``"""
+        return cast(tuple[TypeVar, ...], getattr(self.__origin__, "__parameters__"))
+
+    def _type_var_check(self, args: tuple[type, ...]) -> bool:
+        self._check_generics_resolved()
+        for parameter, self_arg, subclass_arg in zip(
+            self._type_vars(),
+            self.__args__,
+            args,
+            strict=True,
+        ):
+            if not parameter.__contravariant__:
+                # if it's covariant (or invariant in which case check both ways):
+                if not issubform(subclass_arg, self_arg):
+                    return False
+            if not parameter.__covariant__:
+                # if it's contravariant (or invariant in which case check both ways):
+                if not issubform(self_arg, subclass_arg):
+                    return False
+        return True
+
+    def __subclasscheck__(self, subclass: object) -> bool:
+        # could be any random class, check it first
+        if not isinstance(subclass, _ReifiedGenericAlias) or not issubform(
+            subclass.__origin__, self.__origin__
+        ):
+            return False
+        subclass._check_generics_resolved()
+        return self._type_var_check(subclass.__args__)
+
+    def __instancecheck__(self, instance: object) -> bool:
+        # could be any random instance, check it first
+        if not isinstance(instance, self.__origin__) or not isinstance(
+            instance, ReifiedGeneric
+        ):
+            return False
+        return self._type_var_check(instance.__orig_class__.__args__)
+
+
+class OrigClass(Protocol):
+    __args__: tuple[type, ...]
+    __parameters__: tuple[TypeVar, ...]
+
+
+class ReifiedGenericError(TypeError):
+    pass
+
+
+class NotReifiedError(ReifiedGenericError):
+    """raised when a ``ReifiedGeneric`` is instanciated without using a generic alias (ie. the generics can't be reified)"""
+
+
+class UnboundTypeVarError(ReifiedGenericError):
+    """raised when a ``ReifiedGeneric`` is instanciated, ``isinstance`` checked or ``issubclass`` checked with an unbound ``TypeVar``"""
+
+
+class _ReifiedGenericMetaclass(type, OrigClass):
+    def _actual_call(cls, *args: NoReturn, **kwargs: NoReturn) -> object:
+        """the actual  ``__call__`` method for the generic alias's ``__origin__``"""
+        return cast(object, super().__call__(*args, **kwargs))
+
+    def __call__(cls, *args: NoReturn, **kwargs: NoReturn) -> object:
+        """a placeholder ``__call__`` method that only gets called if the ``ReifiedGeneric`` being instanciated isn't a ``_ReifiedGenericAlias``"""
+        raise NotReifiedError(
```

Comment:
So this is only raised when no type parameters are supplied, what about the scenario where 1 / 2 parameters are supplied? Should we check that the correct number are passed?

Reply:
i didn't bother adding runtime checks for stuff that mypy would always catch first. do u think i should?

Reply:
yea i probably should

Reply:
`Generic` already had a check for this so i just added a test to make sure `ReifiedGeneric` still inherits that behavior